### ISSUE TITLE
twisted.spread.banana check for INT vs. LONGINT will fail on some 64-bit platforms

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11057]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#1514|
|Type|defect|
|Created|2006-02-28 16:45:40Z|

The code checks using isinstance(o, int), but the protocol says LONGINT is for >32bit, and ints are 64-bit on some platforms.

Attachments:

* [int64.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/db8/db8dc2b7a3bb754178d96fe9b4e434d0c017825b/bd0ff39fb96bbdc50e675ceed37a60352af46127.diff) (2510 bytes) - added by glyph on 2006-05-30 03:01:51Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1514 1514
type__defect defect
reporter__itamarst itamarst
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1141145140000000 1141145140000000
changetime__1167692277000000 1167692277000000
version__ 
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

I think this patch should fix your issue.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

I don't like testSizedIntegerTypes, because long/int keep getting closer in python, I wouldn't be surprised if long(1) returned an int in python 2.6.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @glyph|
|-|-|

I don't consider python2.6 compatibility a blocker to merging a bugfix for the versions of python that actually exist. This patch exists, it passes the tests that exist, and we have buildbot for the day when python2.6alpha is released.

Please merge this change.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `closed`|
|-|-|

(In [17652]) Allow large numbers to be encoded and decoded by banana on 64-bit hosts.

Fixes [#11057](https://github.com/twisted/twisted/issues/11057)

Author: glyph

Reviewer: itamarst

This fixes a long-standing 64-bit bug in banana.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

James, FWIW, I agree that the test is kind of lame, but at least it provides some coverage.  When the 64-bit buildslave is around more reliably I think we'll probably have more attention devoted to correctness in this area.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

(In [17751]) Revert r17652 - test failures

Refs [#11057](https://github.com/twisted/twisted/issues/11057)

The new test added in this revision fails with Canana on 64 bit platforms.  Canana
should probably just be deleted altogether.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

Probably.  Is it even built by default any more?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @jyknight<br>@glyph set status to `new`|
|-|-|

Seeing that I don't actually have any 64bit OSes any more, I'm probably not the right person to test this and re-merge.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

Test _what_? I don't see anything changed?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Presumably, test some Canana changes that no one has tried to make yet?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

To clarify: canana should be removed, and the fix (which has been reverted) be reinstated.  I can do both of these things, but the only mechanism I have to test is running via the buildbot, which is a pain.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_14"></a>|@radix commented|
|-|-|

UNPRIORITIZED (as release critical)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

Hey look!  Canana is gone.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun removed owner|
|-|-|

I tried applying the patch but it conflicted in both places.  As I was resolving the conflicts in the test module I noticed the test was somewhat incomplete.  I rewrite it to cover several values on each side of each boundary.  As I was about to resolve the conflict in banana.py I noticed the problem had already been fixed by another recent banana branch.

So this is just test changes now, in banana-longint-1514

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_17"></a>|@jyknight set owner to @exarkun|
|-|-|

Looks good. Test passes on both x86-64 and PPC-32, so I concur that the problem was fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set status to `closed`|
|-|-|

(In [19179]) Merge banana-longint-1514

Author: exarkun
Reviewer: jknight
Fixes [#11057](https://github.com/twisted/twisted/issues/11057)

Add test coverage for the behavior of banana int serialization near the
boundary between small integer tokens and large integer tokens, for both
positive and negatives.  No actual fixes to banana, since they seem to
have been made already.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
